### PR TITLE
Fix the global composite operation usage with drawImage

### LINKED ISSUE #416: Bug with Global Composite Operations
The result obtained is different from standard canvas when using the global composite operation 'destination-in' as follow:

```
context.drawImage(imageA, x, y);
context.globalCompositeOperation = 'destination-in';
context.save();
context.translate(imageB.width / 2, imageA.height / 2);
context.rotate(Math.PI / 4);
context.scale(0.5);
context.translate(-imageB.width / 2, -imageA.height / 2);
context.drawImage(imageB, x, y);
context.restore();
```

The global composition 'destination-in' should let ImageA be masked with imageB. But if the bounding box of imageB after transformation is smaller than imageA, then every pixel of imageA outside of this bounding box is displayed, although it should not. But inside this bounding box, the behaviour is correct and imageA is displayed correctly.

It looks like, at least for this global composite operation, the surface that is being considered is the bounding box of the image being drawn although it should be the entire canvas.

### LINKED ISSUE #1167: Shadow applied to cropped image is wrong
I am having some issues when applying a shadow to a cropped image. The shadow is applied correctly when running similar code inside a browser.
```
const fs = require('fs');
const {createCanvas, loadImage} = require('canvas');
const canvas = createCanvas(1200, 500);
const ctx = canvas.getContext('2d');

loadImage('lime-cat.jpg').then((image) => {
  ctx.shadowColor = '#000';
  ctx.shadowBlur = 20;
  ctx.shadowOffsetX = 4;
  ctx.shadowOffsetY = 4;

  // no crop
  // elementToDraw, cropX, cropY, width, height, x, y, width, height
  ctx.drawImage(image, 0, 0, 480, 408, 10, 46, 480, 408);

  // cropped
  ctx.drawImage(image, 100, 100, 380, 308, 760, 126, 380, 308);

  fs.writeFileSync('out.png', canvas.toBuffer());
});
```
out

## Your Environment
* Version of node-canvas (e.g. 1.4.0): canvas@2.0.0-alpha.12
* Environment (e.g. node 4.2.0 on Mac OS X 10.8): node 8.9.4 on Mac OS X 10.13.2

### LINKED ISSUE #1227: Global composite operation with drawImage behaving differently from browser
## Issue
Run the following code in browser or in node-canvas, output differ.

## Steps to Reproduce
```js
const { createCanvas } = require('canvas')
const canvas = createCanvas(400, 400)
const canvas2 = createCanvas(400, 400)
const ctx = canvas.getContext('2d')
const ctx2 = canvas2.getContext('2d')
const fs = require('fs');
const path = require('path');

// Write "Awesome!"
ctx.fillStyle = 'red'
ctx.translate(200, 200);
ctx.beginPath();
ctx.moveTo(-200, -200);
ctx.lineTo(200, -200);
ctx.lineTo(200, 200);
ctx.lineTo(-200, 200);
ctx.lineTo(-200, -200);
ctx.fill();

ctx2.fillStyle = 'black'
ctx2.beginPath();
ctx2.arc(
  200,
  200,
  200,
  0,
  2 * Math.PI, false);
ctx2.fill();

ctx.globalCompositeOperation = 'destination-in';
ctx.drawImage(canvas2, -50, -50);
canvas.createPNGStream().pipe(fs.createWriteStream(path.join(__dirname, 'test.png')))
```
output:
image

Browser version:
http://jsfiddle.net/asturur/hb1o486a/1/

output:
image

The global composite operation effect seems limited to the source graphic extension

## Your Environment
* Version of node-canvas LATEST or NEXT, both
* Environment ( node 8 on Mac OS X 10.13.6)

### PR DESCRIPTION
**Pull request by asturur**
This PR changes the way node-canvas handle the 9 argument version of drawImage.

Before:
The difference between sw and dw, and sh and dh, created a scaling factor.
A clipping area was used to copy just a part of the image, that was also copied as a repeating pattern.

Now:
The drawImage code uses more frequently a separate surface.
A new surface as big as dw * dh is created, and just the right part of the image is copied over there.
Magnified if necessary.
When this magnified or minified copy is created, the main context is modified and the new piece of the image to draw is in the right size and position.

copies of the object are then used to cast shadow and are drawn without the repeating logic.

failing test:
image

After the changes:
image

close #416 
close #1227 
close #1167

### CODE REVIEW

**Comment by asturur** [comment]
well this breaks the copy on the same canvas. Investigating.

**Comment by asturur** [comment]
i wonder how the commit before was green. Shouldn't have been.

**Comment by asturur** [comment]
Also added a visual test for the image crop and shadow issue #1167 

image

**Comment by asturur** [comment]
@LinusU @zbjornson , excluded the extra files i leaked from the other pr, do you have any comments on the changes on the drawImage functionality?

i realized i have to move the tests as failing tests in the other pr and leave here just the renderingContext changes

**Comment by asturur** [comment]
Removed all the extra files.
Ready for review

File: src/CanvasRenderingContext2d.cc [hunk 1]
```diff
@@ -1206,54 +1220,29 @@ NAN_METHOD(Context2d::DrawImage) {
...
-  cairo_rectangle(ctx, dx, dy, dw, dh);
-  cairo_clip(ctx);
-  context->restorePath();
-
   // Paint
-  cairo_set_source_surface(ctx, surface, dx - sx, dy - sy);
+  cairo_set_source_surface(ctx, surface, dx, dy);
   cairo_pattern_set_filter(cairo_get_source(ctx), context->state->patternQuality);
-  cairo_pattern_set_extend(cairo_get_source(ctx), CAIRO_EXTEND_REFLECT);
+  cairo_pattern_set_extend(cairo_get_source(ctx), CAIRO_EXTEND_NONE);
```
Comment (zbjornson) [inline src/CanvasRenderingContext2d.cc]: Ditto.

**Comment by asturur** [comment]
screen shot 2018-09-02 at 10 41 11 am

Here it is! extending JUST the clipped canvas, the extra surface is good enough. Extend will eventually clip just the pixel that is not fully drawn while not replicating the the texture on all the paint operation.

File: src/CanvasRenderingContext2d.cc (hunk 1 above)
Comment (asturur) [reply src/CanvasRenderingContext2d.cc]: This has to be NONE, the previous if fine with REFLECT

**Comment by asturur** [comment]
This code changes swap the clip operation with a canvas that do hard clipping.
For hard clipping i mean that there are no pixels outside the area we want to draw because the canvas is small and so they fall in nothing.

The removal of the clip operation allow the global composite to be extended in all the destination.

File: src/CanvasRenderingContext2d.cc [hunk 2]
```diff
@@ -1178,13 +1180,25 @@ NAN_METHOD(Context2d::DrawImage) {
...
-    dh /= fy;
+  bool sameCanvas = surface == context->canvas()->surface();
+  bool needsExtraSurface = sameCanvas || needCut || needScale;
+  cairo_surface_t *surfTemp = NULL;
+  cairo_t *ctxTemp = NULL;
+
+  if (needsExtraSurface) {
+    surfTemp = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, dw, dh);
+    ctxTemp = cairo_create(surfTemp);
+    cairo_save(ctxTemp);
```
Comment (zbjornson) [inline src/CanvasRenderingContext2d.cc]: I don't think you need this `cairo_save` and the `cairo_restore` 6 lines down, do you?

File: test/public/tests.js [hunk 3]
```diff
@@ -1220,7 +1220,7 @@ tests['known bug #416'] = function (ctx, done) {
       ctx.save()
       ctx.translate(img2.width / 2, img1.height / 2)
       ctx.rotate(Math.PI / 4)
-      ctx.scale(0.5)
+      ctx.scale(0.5, 0.5)
```
[the remark this answers is not in the document]
Comment (asturur) [reply test/public/tests.js]: Well it was not exactly doing nothing, it was breaking the browser code.

broken drawImage, broken test (master):
image


fixed drawImage, broken test:
image

both fixed:
image

Would be nice to throw error for missing second parameter in scale
image

File: src/CanvasRenderingContext2d.cc (hunk 2 above)
Comment (asturur) [reply src/CanvasRenderingContext2d.cc]: Yes was a left over from another change, removed.

File: test/public/tests.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (zbjornson) [reply test/public/tests.js]: >  it was breaking the browser code.

Ish -- I don't know what the user in #416 intended it to look like :)

> Would be nice to throw error for missing second parameter in scale

An annoying part of the HTML Canvas spec is that basically all of the methods should silently return if a parameter is invalid/absent/etc.

**Comment by asturur** [comment]
i always do long commit list because i rely on the 'merge squash' button. if you need me to rebase let me know.

File: test/public/tests.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (asturur) [reply test/public/tests.js]: I see from the specs it has to return if arguments are nan or infinite, but not if they are in the wrong number.
Also firefox return an error on wrong arguments number, and firefox is actually picky about specs.

[the remark this answers is not in the document]
Comment (asturur) [reply test/public/tests.js]: image

[the remark this answers is not in the document]
Comment (zbjornson) [reply test/public/tests.js]: Oof, sorry, you're right, and you even posted a screencap of the error :). I was mis-remembering #1129 and #274.

[the remark this answers is not in the document]
Comment (asturur) [reply test/public/tests.js]: I see not throwing error is wide spread to all method, is something that for sure require its own PR + thinking

**Comment by LinusU** [comment]
Looks great thanks!

**Comment by LinusU** [comment]
Maybe was a bit quick on the merge button 😄 

> Maybe change the browser test labelled "known bug #416" to something else since this fixes it?

This would still be nice to address 👍

**Comment by asturur** [comment]
i ll address!
I need also to backport to 1.6.X. Can we have a 1.6.12 release for this?

Il giorno mar 4 set 2018 alle ore 17:59 Linus Unnebäck <
notifications@github.com> ha scritto:

**Comment by LinusU** [comment]
> I need also to backport to 1.6.X. Can we have a 1.6.12 release for this?

Absolutely 👍

**Pull request merged**